modules/build/src/main/scala/scala/build/preprocessing/PreprocessingUtil.scala

```diff
@@ -11,4 +19,35 @@ object PreprocessingUtil {
   def maybeRead(f: os.Path): Either[BuildException, String] =
     if (os.isFile(f)) Right(os.read(f, defaultCharSet))
     else Left(new FileNotFoundException(f))
+
+  def optionsAndLocationFromDirectives(
+    content: String,
+    scopePath: ScopePath,
+    path: Either[String, os.Path],
+    logger: Logger,
+    maybeRecoverOnError: BuildException => Option[BuildException] = e => Some(e),
+    allowRestrictedFeatures: Boolean
+  ): Either[
+    BuildException,
+    (DirectivesProcessorOutput[BuildOptions], Option[DirectivesLocation])
+  ] = either {
+    val ExtractedDirectives(_, directives0, dirsLocation) =
+      value(from(
+        content.toCharArray,
+        path,
+        logger,
+        Array(UsingDirectiveKind.PlainComment, UsingDirectiveKind.SpecialComment),
+        scopePath,
+        maybeRecoverOnError
+      ))
+    val updatedOptions = value(DirectivesProcessor.process(
```

Comment:
Do I understand correctly that now we process the directives twice for each file? I also see, that `DirectivesProcessor.process` is executed in `ScalaPreprocessor:352`

Reply:
You have right!